sensor: added driver for MCP9808 temperature sensor

alittlebrighter opened the pull request:
I implemented this driver based off of documentation found here: [MCP9808 Datasheet](http://ww1.microchip.com/downloads/en/DeviceDoc/25095A.pdf)
It mostly works but seems to hang after trying to write the the configuration register multiple times.  Could anyone point me to how best to debug where things might be hanging in the I2C interface?  Aside from that issue, does anything need changed in my driver code?

tve commented on sensor/mcp9808/mcp9808.go:
```diff
@@ -0,0 +1,484 @@
+		AboveCritical: !(temp&(1<<15) == 0),
+		AboveUpper:    !(temp&(1<<14) == 0),
+		BelowLower:    !(temp&(1<<13) == 0)}
+	tempResult.CelsiusDeg = wholeNum + (float64(fraction) / 10000.0)
```
This looks wrong to me. Why is it not simply `float64(temp & 0xFFF) / 16` ?

tve commented on sensor/mcp9808/mcp9808.go:
```diff
@@ -0,0 +1,484 @@
+}
+
+// readTempC reads from the ambient temperature register and returns the current temperature value in celsius
+func (d *MCP9808) readTempC(reg byte) (float64, error) {
```
This has a lot of code duplicated with AmbientTemp, why not factor out the common code? E.g. AmbientTemp could call readTempC and then add the info in the extra bits...

tve commented on sensor/mcp9808/mcp9808.go:
```diff
@@ -0,0 +1,484 @@
+		newTemp *= -1
+		signBit = 0x1000
+	}
+	newTempWord := uint16(newTemp)*15 + uint16(newTemp*16) + signBit
```
This is incorrect. it should just be `uint16(newTemp*16) & 0xfff`

tve commented on sensor/mcp9808/mcp9808.go:
```diff
@@ -0,0 +1,484 @@
+
+// SetCriticalTemp when the temperature goes above the set value the alert will be
+// triggered if enabled.
+func (d *MCP9808) SetCriticalTemp(newTemp float64) error {
```
Why is this actually writing to the device when most other Set functions don't

alittlebrighter replied:
Most of the other Set functions are writing to the config which I thought would be cleaner to make edits locally and then once you are satisfied with your config changes you can write them to the device.

tve commented on sensor/mcp9808/mcp9808.go:
```diff
@@ -0,0 +1,484 @@
+	defer d.cmu.Unlock()
+
+	d.SetWindowTempLock(false)
+	if _, err := d.WriteConfig(); err != nil {
```
This is going to deadlock unless I'm missing something 'cause WriteConfig tries to acquire the lock you already hold. You need a writeConfig that doesn't lock that you can call here.

tve commented on samples/mcp9808.go:
```diff
@@ -0,0 +1,112 @@
+
+	alert := rpio.Pin(4)
+	alert.Input()
+	alert.PullDown()
```
This should be pull-up since the alert output is open-drain.

alittlebrighter replied:
I was trying everything and it wasn't working.  I'll have to do some more digging on the alert pin.

tve replied:
The alert pin needs a pull-up, most likely you need an external pull-up, typical values would be 2k-10k. I don't see any recommendation in the datasheet.

tve commented:
Thanks for the PR! Looks nicely done overall :-). I think there are some issue I pointed out and it would be better IMHO if all functions went straight to the device as opposed to making some return info from memory.

tve commented on sensor/mcp9808/mcp9808.go:
```diff
@@ -0,0 +1,484 @@
+// MCP9808 represents a MCP9808 temperature sensor.
+type MCP9808 struct {
+	Bus  embd.I2CBus
+	Poll time.Duration
```
Poll is unused.

alittlebrighter commented on samples/mcp9808.go:
```diff
@@ -0,0 +1,112 @@
+	"github.com/kidoman/embd"
+	_ "github.com/kidoman/embd/host/rpi"
+	"github.com/kidoman/embd/sensor/mcp9808"
+	"github.com/stianeikeland/go-rpio"
```
I apologize, I was getting a weird "Not Implemented" error with embd when I was trying to do a read on the pin and forgot to go back and trace what I was doing wrong.  I'll get rid of that dependency.

tve replied:
If you're having issues with the gpio, let me know...

alittlebrighter commented on samples/mcp9808.go:
```diff
@@ -0,0 +1,112 @@
+	defer rpio.Close()
+	defer embd.CloseGPIO()
+
+	alert := rpio.Pin(4)
```
No, you are correct I need to get rid of rpio in the example.

alittlebrighter commented on sensor/mcp9808/mcp9808.go:
```diff
@@ -0,0 +1,484 @@
+	address = 0x18
+
+	// Register addresses.
+	regConfig       = 0x01
```
Good call, I had just copied those values out of a Python library for this sensor.

tve replied:
on second thought, since the values are not inherently sequential, maybe what you have is better...

alittlebrighter commented on sensor/mcp9808/mcp9808.go:
```diff
@@ -0,0 +1,484 @@
+	Bus  embd.I2CBus
+	Poll time.Duration
+
+	cmu sync.RWMutex
```
I was just trying to match code I found in another sensor driver.

tve replied:
I think the best way to handle the locking is a single simple sync.Mutex. Most or all exported (capitalized) functions start by acquiring the mutex and then deferring an unlock. Few or none of the internal functions touch the mutex. This means that in general an exported function cannot call another exported function 'cause you get a deadlock trying to acquire the mutex twice. In such as situation, you split the called function in two: an internal function with the guts and an exported one that is just a wrapper to acquire the mutex. This design pattern is pretty common in Go in general.

alittlebrighter commented:
Thanks for the review.  Most of my work doesn't involve bit-twiddling so I appreciate you taking the time to correct some of my missteps in that regard.  I'll get those changes made and see what you think.  

In regards to the Set functions on config parameters since it didn't seem like I could write just a single bit to the device I thought it would be easier to make all of your config changes with a local copy and write the entire config word to the device as a separate step saving time making writes over the I2C bus.  If that doesn't make sense I can change it so all of the getters and setters go straight to the device.

tve commented:
I'm not sure you caught the likely cause of your deadlock (see comment for line 442).

I do think that making the set function go straight to the device would be more intuitive. As far as I can tell, everything pretty much gets set once and that's it, it's not like those changes happen frequently or are in some critical section. For this reason I would prioritize ease of use over performance.

alittlebrighter commented on sensor/mcp9808/mcp9808.go:
```diff
@@ -0,0 +1,484 @@
+		return 0, 0, err
+	}
+
+	return uint8(devIDRev >> 8), uint8(devIDRev & 0xFF), nil
```
How do I get a slice of an int?

tve replied:
you would something like:
```
devIDRev := []byte{0,0} // or make([]byte, 2)
err := d.Bus.ReadFromReg(address, regDeviceID, devIDRev) // read 2 bytes
...
return devIDRev[0], devIDRev[1], nil
```

alittlebrighter commented:
I had a crazy weekend but I believe I've made all of the fixes you suggested and fixed an issue with temperature reading/writing that I had missed in the documentation.  Let me know if this version looks better.

I still can't get the interrupt to work although the comparators work just fine.

tve commented on sensor/mcp9808/mcp9808.go:
```diff
@@ -318,65 +308,77 @@ type Temperature struct {
+}
 
-	wholeTempC := temp &^ 0xF000
+func convertWordToTempC(temp uint16) float64 {
```
This really should be just `float64(temp&0xFFF) / 16.0`

alittlebrighter replied:
Yes, I realized about the two's complement format as well hence the change for how I handle the fraction `fraction := float64(temp&0xF) * .0625`
I'm not sure how shifting the bit string back and forth like that is simpler or is more accurate than what I have.  Especially since the results I'm getting from the sensor are accurate and after the changes from your first pass I believe are much simpler as well.

tve replied:
Nope, you're not doing the two's complement correctly. Here's a simple test case:
```
> cat twos.go
package main

import "fmt"

func main() {
        var temp uint16 = 0x1fff // temperature -1
        // broken code
        wholeNum := float64(temp&0xFF0) / 16.0
        fraction := float64(temp&0xF) * .0625

        tempRead := wholeNum + fraction

        if !(temp&(1<<12) == 0) { // read sign bit
                tempRead *= -1
        }
        fmt.Printf("register value %#x -> %fC\n", temp, tempRead)
        // correct code
        tFloat := float64(int16(temp<<3)>>3) / 16
        fmt.Printf("register value %#x -> %fC\n", temp, tFloat)

}
> go run twos.go
register value 0x1fff -> -255.937500C
register value 0x1fff -> -0.062500C
```
As you can see my code is producing the correct value for a two's complement number while your's isn't. The key is to understand that the `>>3` does the necessary sign-extension.

tve commented on sensor/mcp9808/mcp9808.go:
```diff
@@ -318,65 +308,77 @@ type Temperature struct {
+	return tempRead
 }
 
 func (d *MCP9808) setTemp(reg byte, newTemp float64) error {
```
The number format conversion here is incorrect and can be vastly simplified. Something like:
```
intTemp := uint16(newTemp*16+0.5) & 0x1FFF
```
that's all it takes.

alittlebrighter replied:
Which method are you referring to here?  The diff you've selected is inside of the convertWordToTempC function but the line you have in the comment looks like you're referring to the setTemp function.  I've watched the values back and forth as I've set them and both of those functions work as intended (they return the right values).  The setTemp has the switch statement because each temperature that you set can be set to .25 precision.  Your version appears to round to the nearest whole number which is fine but the sensor allows for more precision.  It also doesn't account for the sign bit

tve replied:
my version is a single line and does not round to a whole number at all! The +0.5 rounds to the closest 1/16th, which is the precision of the sensor. But it has some not-so-great properties for negative numbers (rounds up). Test case:
```
> cat cnv.go
package main

import "fmt"

func main() {
        tFloat := -0.062500
        intTemp := int16(tFloat*16) & 0x1fff
        fmt.Printf("register value %#x <- %fC\n", intTemp, tFloat)
}
> go run cnv.go
register value 0x1fff <- -0.062500C
```

alittlebrighter commented:
I'll have to study that code some more and read up on two's complement but thanks for the examples.  It should be good to go now.

tve commented:
did the "hangs in the I2C bus" issue get fixed?

alittlebrighter commented:
Yes, that was just my own stupidity with the locking.  Also, I just fixed the rounding issue.  Per section 5.1.2 in the documentation the limit registers are only precise to a quarter of a degree Celsius hence I changed the mask to 0x1ffc and per some testing [here](https://play.golang.org/p/rFJNgJXKbO) adding two gets the closest to accurate rounding for both positive (exact) and negative (off by a little) values.

tve commented:
LGTM, are there more change you'd like to make?

alittlebrighter commented:
Those last two commits are it.

alittlebrighter commented:
Is there anything else I need to do on this?
